File: playwright-java/src/test/java/testtracks/support/ScenarioState.java
package testtracks.support;

import com.microsoft.playwright.APIRequestContext;
import com.microsoft.playwright.Page;
import java.net.URI;
import testtracks.cleanup.CleanupRegistry;
import testtracks.configuration.EnvironmentDataStore;
import testtracks.configuration.EnvironmentTargets;
import testtracks.configuration.TestRunConfiguration;
import testtracks.configuration.TestSettings;
import testtracks.data.TestData;
import testtracks.diagnostics.ApiEvidence;
import testtracks.runtime.BrowserSession;
import testtracks.runtime.ScenarioResources;

public final class ScenarioState {
  private String siteTargetName;

  public ScenarioState() {
    var testRun = TestRunConfiguration.current();
    scenarioId = TestData.newScenarioId();
    environmentName = testRun.environmentName();
    data = testRun.data();
    targets = data.load("all-targets", EnvironmentTargets.class);
    settings = testRun.settings();
  }

  private final String scenarioId;
  private final String environmentName;
  private final EnvironmentDataStore data;
  private final EnvironmentTargets targets;
  private final TestSettings settings;
  private final CleanupRegistry cleanup = new CleanupRegistry();
  private final ScenarioResources resources = new ScenarioResources();
  private final ApiEvidence apiEvidence = new ApiEvidence();

  public String scenarioId() {
    return scenarioId;
  }

  public String environmentName() {
    return environmentName;
  }

  public EnvironmentDataStore data() {
    return data;
  }

  public EnvironmentTargets targets() {
    return targets;
  }

  public TestSettings settings() {
    return settings;
  }

  public CleanupRegistry cleanup() {
    return cleanup;
  }

  public ScenarioResources resources() {
    return resources;
  }

  public ApiEvidence apiEvidence() {
    return apiEvidence;
  }

  public <TApi> TApi useApi(String apiName, ApiFactory<TApi> factory) {
    var apiBaseUrl = URI.create(targets.api(apiName).baseUrl());
    var session = resources.getOrCreateApi(apiName, apiBaseUrl);
    return factory.create(settings, session.request(), apiEvidence);
  }

  public Page usePage(String siteName) {
    if (siteName.equals(siteTargetName) && resources.browser() != null) {
      return resources.browser().page();
    }

    if (resources.browser() != null) {
      throw new IllegalStateException(
          "This scenario is already using UI site target '"
              + siteTargetName
              + "'. Multiple browser sessions are not configured in this example state.");
    }

    var siteBaseUrl = URI.create(targets.site(siteName).baseUrl());
    resources.browser(BrowserSession.create(settings, siteBaseUrl, scenarioId));
    siteTargetName = siteName;

    return resources.browser().page();
  }

  @FunctionalInterface
  public interface ApiFactory<TApi> {
    TApi create(TestSettings settings, APIRequestContext request, ApiEvidence evidence);
  }
}

File: playwright-csharp/tests/Support/ScenarioState.cs
using Microsoft.Playwright;
using TestTracks.Playwright.CSharp.Cleanup;
using TestTracks.Playwright.CSharp.Configuration;
using TestTracks.Playwright.CSharp.Diagnostics;
using TestTracks.Playwright.CSharp.Runtime;
using TestTracks.Playwright.CSharp.Specs.Configuration;
using TestTracks.Playwright.CSharp.Specs.Data;

namespace TestTracks.Playwright.CSharp.Specs.Support;

public sealed class ScenarioState
{
    private string? _siteTargetName;

    public ScenarioState()
    {
        var testRun = TestRunConfiguration.Current;

        ScenarioId = TestData.NewScenarioId();
        EnvironmentName = testRun.EnvironmentName;
        Data = testRun.Data;
        Targets = Data.Load<EnvironmentTargets>("all-targets");
        Settings = testRun.Settings;
        Cleanup = new CleanupRegistry();
        ApiEvidence = new ApiEvidence();
    }

    public string ScenarioId { get; }

    public string EnvironmentName { get; }

    public EnvironmentDataStore Data { get; }

    public EnvironmentTargets Targets { get; }

    public TestSettings Settings { get; }

    public CleanupRegistry Cleanup { get; }

    public ScenarioResources Resources { get; } = new();

    public ApiEvidence ApiEvidence { get; }

    public async Task<TApi> UseApiAsync<TApi>(
        string apiName,
        Func<TestSettings, IAPIRequestContext, ApiEvidence, TApi> factory)
    {
        var apiBaseUrl = new Uri(Targets.Api(apiName).BaseUrl);
        var session = await Resources.GetOrCreateApiAsync(apiName, apiBaseUrl);
        return factory(Settings, session.Request, ApiEvidence);
    }

    public async Task<IPage> UsePageAsync(string siteName)
    {
        if (_siteTargetName == siteName && Resources.Browser is not null)
        {
            return Resources.Browser.Page;
        }

        if (Resources.Browser is not null)
        {
            throw new InvalidOperationException(
                $"This scenario is already using UI site target '{_siteTargetName}'. Multiple browser sessions are not configured in this example state.");
        }

        var siteBaseUrl = new Uri(Targets.Site(siteName).BaseUrl);
        Resources.Browser = await BrowserSession.CreateAsync(Settings, siteBaseUrl, ScenarioId);
        _siteTargetName = siteName;

        return Resources.Browser.Page;
    }
}
